pkg/controller/registrysyncer/registrysyncer.go

```diff
@@ -0,0 +1,437 @@
+func (r *reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithField("request", req.String())
+	log.Info("Starting reconciliation")
+	err := r.reconcile(req, log)
+	if err != nil {
+		log.WithError(err).Error("Reconciliation failed")
+	} else {
+		log.Info("Finished reconciliation")
+	}
+	return reconcile.Result{}, controllerutil.SwallowIfTerminal(err)
+}
+
+func (r *reconciler) reconcile(req reconcile.Request, log *logrus.Entry) error {
+	cluster, decoded, err := decodeRequest(req)
+	if err != nil {
+		return fmt.Errorf("failed to decode request %s: %w", req, err)
+	}
+
+	// Propagate the cluster field back up
+	*log = *log.WithField("cluster", cluster)
+
+	isTags := map[string]*imagev1.ImageStreamTag{}
+	for ctx, client := range r.registryClients {
+		imageStreamTag := &imagev1.ImageStreamTag{}
+		if err := client.Get(r.ctx, decoded, imageStreamTag); err != nil {
+			if apierrors.IsNotFound(err) {
+				log.Debug("Source imageStreamTag not found")
+				isTags[ctx] = nil
+				return nil
+			}
+			return fmt.Errorf("failed to get imageStreamTag %s from registry cluster: %w", decoded.String(), err)
+		}
+		isTags[ctx] = imageStreamTag
+	}
+
+	srcContext, err := findNewest(isTags)
+	if err != nil {
+		return fmt.Errorf("failed to find the new isTag from %v: %w", isTags, err)
+	}
+	sourceImageStreamTag := isTags[srcContext]
+
+	imageStreamNameAndTag := strings.Split(decoded.Name, ":")
+	if n := len(imageStreamNameAndTag); n != 2 {
+		return fmt.Errorf("when splitting imagestreamtagname %s by : expected two results, got %d", decoded.Name, n)
+	}
+	imageStreamName, imageTag := imageStreamNameAndTag[0], imageStreamNameAndTag[1]
+	isName := types.NamespacedName{Namespace: decoded.Namespace, Name: imageStreamName}
+	sourceImageStream := &imagev1.ImageStream{}
+	if err := r.registryClients[srcContext].Get(r.ctx, isName, sourceImageStream); err != nil {
+		return fmt.Errorf("failed to get imageStream %s from registry cluster: %w", isName.String(), err)
+	}
+
+	*log = *log.WithField("docker_image_reference", sourceImageStreamTag.Image.DockerImageReference)
+	if isImportForbidden(sourceImageStreamTag.Image.DockerImageReference, r.forbiddenRegistries) {
+		log.Debugf("Import from any cluster in %s is forbidden, ignoring", r.forbiddenRegistries)
+		return nil
+	}
+
+	for ctx, isTag := range isTags {
+		if ctx == srcContext {
+			continue
+		}
+		client := r.registryClients[ctx]
+		if err := client.Get(r.ctx, types.NamespacedName{Name: decoded.Namespace}, &corev1.Namespace{}); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return fmt.Errorf("failed to check if namespace %s exists: %w", decoded.Namespace, err)
+			}
+			if err := client.Create(r.ctx, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: decoded.Namespace}}); err != nil && !apierrors.IsAlreadyExists(err) {
+				return fmt.Errorf("failed to create namespace %s: %w", decoded.Namespace, err)
+			}
+		}
+
+		imageStreamImport := &imagev1.ImageStreamImport{
```

Comment:
We need to upsert the ImagePullSeret first

Reply:
I moved it up here because I wanted to log it at dry run without creating new objects.
Moved it back now because dryRun is recognized by the controller as pointed out https://github.com/openshift/ci-tools/pull/1376#discussion_r517650793